src/modules/module_25200.c

```diff
@@ -0,0 +1,318 @@
+                   | TOKEN_ATTR_VERIFY_HEX;
+
+  const int rc_tokenizer = input_tokenizer ((const u8 *) line_buf, line_len, &token);
+
+  if (rc_tokenizer != PARSER_OK) return (rc_tokenizer);
+
+  // packet number
+  const u8 *packet_number_pos = token.buf[1];
+  const int packet_number_len = token.len[1];
+
+  memset (snmpv3->packet_number, 0, sizeof (snmpv3->packet_number));
+
+  strncpy ((char *) snmpv3->packet_number, (char *) packet_number_pos, packet_number_len);
+
+  // salt
+
+  const u8 *salt_pos = token.buf[2];
+  const int salt_len = token.len[2];
+
+  u8 *salt_ptr = (u8 *) snmpv3->salt_buf;
+
+  int i;
+  int j;
+
+  for (i = 0, j = 0; i < salt_len; i += 2, j += 1)
+  {
+    salt_ptr[j] = hex_to_u8 (salt_pos + i);
+  }
+
+  for (i = 0; i < salt_len / 4; i++)
+  {
+    snmpv3->salt_buf[i] = byte_swap_32 (snmpv3->salt_buf[i]);
+  }
+
+  snmpv3->salt_len = salt_len / 2;
+
+  salt->salt_iter = 16384 - 1;
+
+  // handle unique salts detection
+
+  sha1_ctx_t sha1_ctx;
+
+  sha1_init   (&sha1_ctx);
+  sha1_update (&sha1_ctx, snmpv3->salt_buf, snmpv3->salt_len);
+  sha1_final  (&sha1_ctx);
+
+  // store sha1(snmpv3->salt_buf) in salt_buf
+
+  memcpy (salt->salt_buf, sha1_ctx.h, 20);
+
+  // engineid
+
+  const u8 *engineID_pos = token.buf[3];
+  const int engineID_len = token.len[3];
+
+  u8 *engineID_ptr = (u8 *) snmpv3->engineID_buf;
+
+  for (i = 0, j = 0; i < engineID_len; i += 2, j += 1)
+  {
+    engineID_ptr[j] = hex_to_u8 (engineID_pos + i);
+  }
+
+  snmpv3->engineID_len = engineID_len / 2;
+
+  // digest
+
+  const u8 *hash_pos = token.buf[4];
+
+  digest[0] = hex_to_u32 (hash_pos +  0);
+  digest[1] = hex_to_u32 (hash_pos +  8);
+  digest[2] = hex_to_u32 (hash_pos + 16);
+
+  digest[0] = byte_swap_32 (digest[0]);
+  digest[1] = byte_swap_32 (digest[1]);
+  digest[2] = byte_swap_32 (digest[2]);
+
+  digest[3] = 0;
+
+  return (PARSER_OK);
+}
+
+int module_hash_encode (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const void *digest_buf, MAYBE_UNUSED const salt_t *salt, MAYBE_UNUSED const void *esalt_buf, MAYBE_UNUSED const void *hook_salt_buf, MAYBE_UNUSED const hashinfo_t *hash_info, char *line_buf, MAYBE_UNUSED const int line_size)
+{
+  const u32 *digest = (const u32 *) digest_buf;
+
+  snmpv3_t *snmpv3 = (snmpv3_t *) esalt_buf;
+
+  int line_len = snprintf (line_buf, 10 + strlen ((char *) snmpv3->packet_number) + 1 + 1, "%s%s$", SIGNATURE_SNMPV3, snmpv3->packet_number);
+
+  uint i;
+
+  u32 salt_buf_32[SNMPV3_SALT_MAX_BIN] = { 0 };
+
+  for (i = 0; i < SNMPV3_SALT_MAX_BIN; i++) salt_buf_32[i] = byte_swap_32 (snmpv3->salt_buf[i]);
+
+  const u8 *salt_buf_ptr = (u8 *) salt_buf_32;
+
+  for (i = 0; i < snmpv3->salt_len; i++)
+  {
+    line_len += snprintf (line_buf+line_len, 3, "%02x", salt_buf_ptr[i]);
+  }
+
+  line_len += snprintf (line_buf+line_len, 2, "$");
+
+  for (i = 0; i < snmpv3->engineID_len; i++)
+  {
+    line_len += snprintf (line_buf+line_len, 3, "%02x", snmpv3->engineID_buf[i]);
```

Comment:
We hex_decode(), why not use it?

Reply:
Hi, this is hex_encode() but yes, I updated the code to use both hex_encode and hex_decode. Thank you!